### ISSUE TITLE
Use bvhtree for forward+ lamp culling 

### ISSUE THREAD

**Issue by BluePrintRandom** [UPBGE/upbge#776]
It is possible to build a bvhtree each Frame of all active lamps, and use this tree to discard lamps in the shader, this will improve rendering time in almost all cases, without using a compute shader, making it good for compatibility as well as performance. 

Thank you gentlemen for your efforts. 
I can't wait to test out master when I am resettled from the fires.

**Comment by panzergame** [comment]
I'm really not an expert on BVH but i'm still looking for a way to create a tree quickly every frame or dynamically updating one. This is common for many area such culling, occlusion…

@BluePrintRandom : Do you known some ?

**Comment by BluePrintRandom** [comment]
Apparently there are a few libs that make bvhtree quickly
Some by using openCL/Cuda and others that use half percision and a hiarchy.

https://github.com/brandonpelfrey/Fast-BVH/blob/master/README

We have a solution built into blender but I don't think it is threaded or
using the gpu at all.

Maybe we accelerate Bvhtree mathutils lib as well as have a solution that
builds the tree on the gpu for the lamps? (sooner or later?)

**Comment by UnidayStudio** [comment]
Oh maybe only update the tree when a lamp moves/change its radius. It's a good point to create a lamp culling based on the radius. If a lamp+radius isn't inside the camera view, there is no need to include them in the shader calculation.

But a question: Is the lamps really slow? I made a test adding as much point lights as I could and I noticed no framedrop... Even with 300, 400 lights... At some point, blender stops rendering the ilumination (I don't knowwhy, probably I reached the lamp limit supported). But no framedrop...
Same with the spot lights (with shadows disabled). And for these one, I've a number: 62 is the limit. If I add 63 spot lights, blender (and ge) stops rendering the lights, displaying everything black inside the 3D view and everything white inside GE. Again: No framedrop detected...

Another point: Why not move to deferred rendering? This option will solve any of these problems, since the max complexity rendering the light is the amount of pixels on the screen multiplied by the lamps. (Better than forward, which is the max pixels not culled during the rasterization * the lamps.

What is the actuall dificulty of implementing deferred rendering right now?
